### PR TITLE
Support native build on macOS Powered by ARM.

### PR DESCRIPTION
**Pull request by YLinXin**

### CODE REVIEW

File: numpy/distutils/system_info.py [hunk 1]
```diff
@@ -1911,31 +1911,36 @@ def calc_info(self):
...
-                if intel:
+                if arm:
+                    pass
+                elif intel:
                     args.extend(['-msse3'])
                 else:
                     args.extend(['-faltivec'])
                 args.extend([
-                    '-I/System/Library/Frameworks/vecLib.framework/Headers'])
+                    '-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Headers'])
```
Comment (rgommers) [inline numpy/distutils/system_info.py:1931]: This looks like it's going to break things. For me, the path is slightly different: `Versions/A/Frameworks` is `Versions/Frameworks` (no `A/`) for me in macOS 10.14

Also the `MacOSX.sdk` path I'm not sure about. I see:
```
DriverKit19.0.sdk	MacOSX.sdk		MacOSX10.15.sdk
```
there. Does this work for different SDK versions?

Comment (rgommers) [reply numpy/distutils/system_info.py:1931]: Is there an explanation somewhere about why this is the correct path?

File: numpy/distutils/system_info.py [hunk 2]
```diff
@@ -1911,31 +1911,36 @@ def calc_info(self):
...
                'i386' in platform.platform():
                 intel = 1
-            else:
-                intel = 0
             if (os.path.exists('/System/Library/Frameworks'
                               '/Accelerate.framework/') and
                     'accelerate' in libraries):
-                if intel:
+                if arm:
+                    pass
```
Comment (rgommers) [inline numpy/distutils/system_info.py:1925]: Do we need an alternative here? We have Neon intrinsics, maybe they can be enabled?

**Review by rgommers — COMMENTED** [review]
Thanks for working on this @YLinXin!

Just a couple of questions. Also for reference, https://github.com/MacPython/numpy-wheels/pull/104 builds `arm64` wheels that will be released very soon (in 1.21.0, perhaps next week).

File: numpy/distutils/system_info.py (hunk 1 above)
Comment (YLinXin) [reply numpy/distutils/system_info.py:1931]: I've learned a lot about this code section, my project now on python27 so the newest v1.21.0 cannot suit my demand.

Here is my env.
Hardware:
1. Chip:	Apple M1

Software:
1. macOS 11.4(20F71)
2. Xcode 12.5 Build version 12E262 both IDE and command line.

`
ls -alh /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework/Versions/                            
total 0
drwxr-xr-x  4 root  wheel   128B May 30 14:37 .
drwxr-xr-x  7 root  wheel   224B May 30 14:37 ..
drwxr-xr-x  6 root  wheel   192B Mar 16 22:04 A
lrwxr-xr-x  1 root  wheel     1B May 30 14:25 Current -> A

`

`
locate 'vecLib.framework/Headers'|grep -v .h
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/BNNS
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/LinearAlgebra
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/Quadrature
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/Sparse
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/BNNS
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/LinearAlgebra
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/Quadrature
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Headers/Sparse
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Headers
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Headers
/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Headers
/Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Headers

`

Maybe the chaos macOS SDK path leads to this.
I think both the old and the new path cannot solve this build warning on every macOS version.
The wrong header path will take no effect build but error.

File: numpy/distutils/system_info.py (hunk 2 above)
Comment (YLinXin) [reply numpy/distutils/system_info.py:1925]: First I order the three architecture: the future ARM, the now X86, and the past PowerPC.

1. I've found no explicit options on clang to enable Neon intrinsics
```
xxx@MacBook-Pro numpy-1.16.6 % clang --help|grep neon       
xxx@MacBook-Pro numpy-1.16.6 % clang --help|grep simd
  -fopenmp-simd           Emit OpenMP code only for SIMD-based constructs.
xxx@MacBook-Pro numpy-1.16.6 % clang -v
Apple clang version 12.0.5 (clang-1205.0.22.9)
Target: arm64-apple-darwin20.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
maybe this apple clang default on Neon intrinsics, or maybe future clang will supply some explicit options?

gcc simd options is [here](https://gcc.gnu.org/onlinedocs/gcc-11.1.0/gcc/AArch64-Options.html#AArch64-Options)
`
‘simd’
Enable Advanced SIMD instructions. This also enables floating-point instructions. This is on by default for all possible values for options -march and -mcpu. 
`

2. should we ever support the PowerPC nowadays on macOS and the '-faltivec' option will cause warning on build:
`clang: error: the clang compiler does not support 'faltivec', please use -maltivec and include altivec.h explicitly`

[PowerPC cmdoption-clang-maltivec](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-maltivec)

replace '-faltivec' wiht '-maltivec':
[clang](https://gitlab.tu-berlin.de/daniel-schuermann/clang/commit/a776ed1122e48f9ade40c454678226303d8fab52)
[gcc](https://gcc.gnu.org/legacy-ml/gcc-patches/2004-05/msg01327.html)

**Comment by YLinXin** [comment]
The azure pipeline seems so many env problems.

**Comment by rgommers** [comment]
This PR is stale - native macOS arm64 builds are working just fine. So I'll go ahead and close this PR. Thanks for your contribution @YLinXin

**Pull request closed without merging**